File: BST/deleteNode.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
  if (!root) return null;

  if (key < root.val) {
    root.left = deleteNode(root.left, key);
  } else if (key > root.val) {
    root.right = deleteNode(root.right, key);

    // found node
  } else {
    // found node has no left child or subtree
    if (!root.left) {
      return root.right;
      // found node has child or subtree but not right child
    } else if (!root.right) {
      return root.left;
    }

    // found node has both left and right child nodes
    // in this case we can either
    // replace found node with the min value of right subtree OR max value of left subtree
    let minNode = findMin(root.right);
    root.val = minNode.val;

    // recursive re-do the same process
    // this will continue travering down the tree until found found value
    // is a leaf node and is swapped for a null
    root.right = deleteNode(root.right, root.val);
  }
  return root;
}

function findMin(node: TreeNode | null): TreeNode {
  while (node.left !== null) {
    node = node.left;
  }
  return node;
}

File: BST/delete_node.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:    
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if not root: 
            return 
        
        if key > root.val: 
            root.right = self.deleteNode(root.right, key)
        
        if key < root.val: 
            root.left = self.deleteNode(root.left, key)
            
        # we found our value
        else: 
            # scenario: there is no left subtree or left child
            # then return the right child to the parent node
            if not root.left: 
                return root.right
            
            #  scenario it has a left child so we want the max value on the left subtree
            #  to replace the node we want to delete
            else:
                current = root.left
                while current.right: 
                    current = current.right
                
                # replace key val with found max value of left subtree
                root.val = current.val
                
                # replace the max value from the left subtree
                # by recursive do the same process
                root.left = self.deleteNode(root.left, current.val)
        return root